sklearn/calibration.py

```diff
@@ -25,6 +27,368 @@
+
+        y_score = _get_binary_score(self.base_estimator, X, self.method,
+                                    self.pos_label)
+        return self.label_encoder_.inverse_transform(
+            (y_score > self.decision_threshold_).astype(int)
+        )
+
+
+class _CutoffClassifier(object):
+    """Cutoff point selection.
+
+    It assumes that base_estimator has already been fit, and uses the input set
+    of the fit function to select a cutoff point. Note that this class should
+    not be used as an estimator directly. Use the CutoffClassifier with
+    cv="prefit" instead.
+
+    Parameters
+    ----------
+    base_estimator : obj
+        The binary classifier whose decision threshold will be adapted
+        according to the acquired cutoff point. The estimator must have a
+        decision_function or a predict_proba
+
+    strategy : 'roc' or 'f_beta' or 'max_tpr' or 'max_tnr'
+        The method to use for choosing the cutoff point
+
+    method : str or None, optional (default=None)
+        The method to be used for acquiring the score. Can either be
+        "decision_function" or "predict_proba" or None. If None then
+        decision_function will be used first and if not available
+        predict_proba
+
+    beta : float in [0, 1]
+        beta value to be used in case strategy == 'f_beta'
+
+    threshold : float in [0, 1]
+        minimum required value for the true negative rate (specificity) in case
+        strategy 'max_tpr' is used or for the true positive rate (sensitivity)
+        in case method 'max_tnr' is used
+
+    pos_label : object
+        Label considered as positive during the roc_curve construction
+
+    Attributes
+    ----------
+    decision_threshold_ : float
+        Acquired decision threshold for the positive class
+    """
+    def __init__(self, base_estimator, strategy, method, beta, threshold,
+                 pos_label):
+        self.base_estimator = base_estimator
+        self.strategy = strategy
+        self.method = method
+        self.beta = beta
+        self.threshold = threshold
+        self.pos_label = pos_label
+
+    def fit(self, X, y):
+        """Select a decision threshold for the fitted model's positive class
+        using one of the available methods
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            Training data
+
+        y : array-like, shape (n_samples,)
+            Target values
+
+        Returns
+        -------
+        self : object
+            Instance of self
+        """
+        y_score = _get_binary_score(self.base_estimator, X, self.method,
+                                    self.pos_label)
+        if self.strategy == 'f_beta':
+            precision, recall, thresholds = precision_recall_curve(
+                y, y_score, pos_label=self.pos_label
+            )
+            f_beta = ((1 + self.beta**2) * (precision * recall) /
+                      (self.beta**2 * precision + recall))
```

Comment:
I ran some tests on this, and there is a problem due to the fact that `precision_recall_curve` adds a virtual 0 recall 1 precision point at the beginning of the curve, which is not real (there has been some issues about that in scikit-learn, if I understood correctly this point could be useful to plot the curve (I didn't read everything but I'm still not sure why), but needs to be removed when computing metrics)). This would create a virtual point which would have `f1_score=1` (perfect score), so it would mess with the results (plus it would also shift indices of scores by 1 w.r.t their corresponding threshold). 

Also, for some value of the threshold, there could be `precision=0` and `recall=0`, which would give `nan` as threshold, which would be considered as the highest value. Maybe you could put then nan to zero then (not to select them). This type of strategy is done here for instance https://github.com/scikit-learn/scikit-learn/blob/5fd9e03cf3a2e58d3b76dfaf0692437a4d6e1eec/sklearn/metrics/ranking.py#L525

So I guess you could make these changes: 

```diff
-            f_beta = ((1 + self.beta**2) * (precision * recall) /
-                      (self.beta**2 * precision + recall))
+           precision, recall = precision[1:], recall[1:]
+           f_beta = ((1 + beta**2) * (precision * recall) /
+                     (beta**2 * precision + recall))
+           f_beta[np.isnan(f_beta)] = 0
```

Reply:
@wdevazelhes seem like good points; I will need to investigate them - could share your test?

Reply:
also need to share that I am waiting for a response from the core contributors who made the request for the enhancement before putting significant energy to the pr;

Reply:
Btw I made a mistake, in fact it was [:-1] not [1:] (the added virtual point is at the end), and after thinking about it it might not be useful in fact because the added point has precision=1 and recall=0 which would do a beta score of 0 (except if beta=0 in this case it would be nan). As we set then the nans to zero (we still need to do that I think), this point will be ignored.

Reply:
> also need to share that I am waiting for a response from the core contributors who made the request for the enhancement before putting significant energy to the pr;

No pb, I'll leave my comments here, so that you can use them when you'll have your answer by the contributors and keep going with the PR